File: src/consume-line-comment.js
/**
 * @fileoverview
 * Consumes a single-line comment from JSONC input.
 */

/**
 * @typedef {import('./tokenize-jsonc.js').Token} Token
 */

/**
 * Result from consumeLineComment.
 * @typedef {Object} ConsumeResult
 * @property {number} column - Updated column position
 * @property {number} line - Line number (unchanged for line comments)
 * @property {number} pos - Updated position in the string
 * @property {Token} token - The consumed comment token
 */

/**
 * Consumes a single-line comment (// ...).
 * @param {string} jsonc - The input string
 * @param {number} startPos - Starting position (at the first '/')
 * @param {number} len - Length of input
 * @param {number} line - Current line number
 * @param {number} startColumn - Starting column
 * @returns {ConsumeResult}
 */
export function consumeLineComment(jsonc, startPos, len, line, startColumn) {
    let pos = startPos + 2; // skip '//'
    let column = startColumn + 2;
    let value = '//';

    // Consume until newline or end of input.
    while (pos < len) {
        const ch = jsonc[pos];
        if (ch === '\n' || ch === '\r') {
            break;
        }
        value += ch;
        pos += 1;
        column += 1;
    }

    return {
        column,
        line,
        pos,
        token: { category: 'comment', column: startColumn, line, value },
    };
}

File: wasm/jsonc_lexer/src/consume_line_comment.rs
//! Consumes a single-line comment from JSONC input.

use crate::types::{ConsumeResult, Token};

/// Consumes a single-line comment (// ...).
pub fn consume_line_comment(
    bytes: &[u8],
    start_pos: usize,
    len: usize,
    line: u32,
    start_column: u32,
) -> ConsumeResult {
    let mut pos = start_pos + 2; // skip '//'
    let mut column = start_column + 2;
    let mut value = vec![b'/', b'/'];

    // Consume until newline or end of input.
    while pos < len {
        let ch = bytes[pos];
        if ch == b'\n' || ch == b'\r' {
            break;
        }
        value.push(ch);
        pos += 1;
        column += 1;
    }

    ConsumeResult {
        column,
        line,
        pos,
        token: Token {
            category: "comment".to_string(),
            column: start_column,
            line,
            value: String::from_utf8_lossy(&value).to_string(),
        },
    }
}
